### DISCUSSION TITLE
Using fixture instance in UnitTets with AutoMoqDataAttribute

### DISCUSSION THREAD

**Discussion by MGrgov90** [AutoFixture/AutoFixture#1375] [Q&A]
I need fixture instance created in AutoMoqDataAttribure to be used in unit tests. 
So I can register concrete implementation of **IService** interface in my case **Service1**. 

I know that instead of using concrete implementation I should probably mock it so the test for one class does not depends on another. Also there is a way to use it if I create IFixture instance in my unit test class.

The code sample is something what I interested in.

```csharp
public class AutoMoqDataAttribute: AutoDataAttribute
{
    public AutoMoqDataAttribute(): base(() => new Fixture().Customize(new AutoMoqCustomization()))
    {
    }
}
public class WhenUpdatePerson{
    [Theory]
    [AutoMoqData]
    public async Task PersonShouldBeUpdated(IFixture fixture, //INSTANCE THAT I'M INTERESTED IN
    [Frozen] Mock<IRepository<Person>> personRepositoryMock,
    UpdatePersonCommandHandler handler)
    {
        //THIS IS SOMETHING WHAT I WANT TO DO
        fixture.Register<IService>(() => fixture.Create<Service1>());
        // setup personRepositoryMock, call handler method, assert
    }
}

public class UpdatePersonCommandHandler : IRequestHandler<UpdatePersonCommand>
{
    private readonly IRepository<Person> _personRepository;
    private readonly IService service;

    public UpdatePersonCommandHandler(IRepository<Person> personRepository, IService service)
    {
        _personRepository = personRepository;
        this.service = service;
    }
    public async Task<Unit> Handle(UpdatePersonCommand request, CancellationToken cancellationToken)
    {
        var rnd = service.GetRandom();
        //the application logic
        return Unit.Value;
    }
}

public interface IService
{
    int GetRandom();
}

public class Service1 : IService
{
    public int GetRandom() => new Random().Next(0, 100);
}
```

**Comment by GeraldLx** [comment]
The problem with your approach is the timing. When you let AutoFixture create the parameters for your test, this happens before the test method. So your approach using the fixture for your service is a chicken and egg problem. You cannot register anything AFTER the object is created and hope that your changes will happen magically.

The AutoMoqDataAttribute you show, is the basic sample. While not using Moq but NSubstitute I usually have my own AutoDataAttribute with more functionality to provide support for further customization. 

I dont know what you really want to do, but just to make your example work you could re-write your test like this:
```
  public class WhenUpdatePerson
  {
    [Theory]
    [AutoMoqData]
    public async Task PersonShouldBeUpdated(
      [Frozen(Matching.ImplementedInterfaces)] Service1 service,
      [Frozen] IRepository<Person> personRepositoryMock,
      UpdatePersonCommandHandler handler)
    {
      // ....
    }
  }
```

No need to interact on low level with the fixture instance.

**Comment by aivascu** [comment]
Hi @MGrgov90,
If you want to delegate the resolution of an interface to a concrete class, I recommend using `TypeRelay`.

```cs
public class MyServiceCustomization : ICustomization
{
    public void Customize(IFixture fixture)
    {
        fixture.Customizations.Insert(0, new TypeRelay(typeof(IService), typeof(Service1)));
    }
}
```
To use this customization you'll have to create a new attribute that will look like this.
```cs
public class ServicesDataAttribute : AutoDataAttribute
{
    public ServicesDataAttribute()
        : base(() => new Fixture().Customize(
            new CompositeCustomization(
                new AutoMoqCustomization(),
                new MyServiceCustomization())))
    {
    }
}
```
Now when you'll resolve the service interface you should get the concrete implementation.
```cs
[Theory]
[ServicesData]
public void Foo(IService service)
{
    Assert.IsType<Service1>(service);
}
```
